feat(jest-message-util): add support for error causes

brodo opened the pull request:
## Summary

Adds support for [error causes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause).

This fixes #12999

Example output:
```
● Test suite failed to run
    
        Test exception
    
          at Object.<anonymous> (packages/jest-message-util/src/__tests__/messages.test.ts:419:5)
    
        Cause:
         Cause Error
    
              at Object.<anonymous> (packages/jest-message-util/src/__tests__/messages.test.ts:419:41)

      at Object.log (packages/jest-message-util/src/__tests__/messages.test.ts:428:11)
```

I'm sure this can be made prettier, but as a minimal implementation I think it is good enough.

## Test plan

I've added a single unit test that checks if the cause is contained in the generated message. Otherwise I just looked at the output manually.

mrazauskas commented on packages/jest-message-util/src/__tests__/messages.test.ts:
```diff
@@ -413,3 +413,20 @@ it('getTopFrame should return a path for mjs files', () => {
+      noStackTrace: false,
+    },
+  );
+  expect(message).toEqual(expect.stringContaining('Cause'));
```
Perhaps it makes sense to snapshot it:

```suggestion
  expect(message).toMatchSnapshot();
```

mrazauskas commented on packages/jest-message-util/src/index.ts:
```diff
@@ -122,18 +122,20 @@ function warnAboutWrongTestEnvironment(error: string, env: 'jsdom' | 'node') {
 // are executed.
 export const formatExecError = (
-  error: Error | TestResult.SerializableError | string | undefined,
+  error: Error | TestResult.SerializableError | string | undefined | number,
```
Better:

```suggestion
  error: Error | TestResult.SerializableError | string | number | undefined,
```

mrazauskas commented on packages/jest-message-util/src/__tests__/messages.test.ts:
```diff
@@ -413,3 +413,20 @@ it('getTopFrame should return a path for mjs files', () => {
+
+it('should return the error cause if there is one', () => {
+  const error = new Error('Test exception');
+  // We need to do it this way because in older Javascript engines the constructor does not allow supplying a cause.
```
```suggestion
  // TODO pass `cause` to the `Error` constructor when lowest supported Node version is 16.9 and above
```

brodo commented:
@mrazauskas Thanks for the quick review. I did all the changes and added the snapshot file. I also added a link to the Node changelog to where this feature was introduced. (It was indeed 16.9.0).

SimenB commented on packages/jest-message-util/src/index.ts:
```diff
@@ -145,6 +147,27 @@ export const formatExecError = (
+        cause += `${prefix}${error.cause}`;
+      } else if (typeof error.cause === 'number') {
+        cause += `${prefix}${error.cause.toString()}`;
+      } else if (error.cause instanceof Error) {
```
use `types.isNativeError` instead - it works cross realms: https://nodejs.org/api/util.html#utiltypesisnativeerrorvalue

SimenB reviewed (approved):
sweet, thanks!

SimenB commented on packages/jest-message-util/src/index.ts:
```diff
@@ -145,6 +147,27 @@ export const formatExecError = (
+      if (typeof error.cause === 'string') {
+        cause += `${prefix}${error.cause}`;
+      } else if (typeof error.cause === 'number') {
+        cause += `${prefix}${error.cause.toString()}`;
```
do we need the `toString` call?

brodo replied:
Oh, that's right. We don't. 😀
I'll push an updated version soon!

brodo commented:
I implemented all the changes @SimenB requested.

BTW: I don't know if keeping this all in one commit is the best way to do it. I assumed it's the best, so it's only one commit when it's merged but I don't know if it's a good idea for the review process. Would it be better to have several commits for the review process and then squash them before they are merged?

SimenB commented:
We squash on merge anyways, so multiple commits are easier for review. The diff in this PR is quite small though, so no big issue 🙂

GerkinDev commented:
Hey, thank you very much for this ! GG

dubzzz commented:
I might be wrong but it seems that jest-circus already pre-processes the error, leading to this code not to be executed for something like:

```js
function g() {
  throw new Error("boom");
}
function f() {
  try {
    g();
  } catch (err) {
    throw new Error("plouf", { cause: err });
  }
}
it("works", () => {
  f();
});
```

Do I miss something to properly use cause?

SimenB commented:
we should work with causes inside functions as well. Wanna send a PR? 😀

dubzzz commented:
> Wanna send a PR?

Currently trying to check how to do. As errors are extracted from a string in jest-message-util (see formatResultsErrors which uses result.failureMessages, a string version). 😂

SimenB commented:
IIRC it's https://github.com/facebook/jest/blob/6d2632adae0f0fa1fe116d3b475fd9783d0de1b5/packages/jest-circus/src/utils.ts#L438-L439

dubzzz commented:
By the way, I do have a working version of it locally I switched from result.failureMessage to result.failureDetails, but I'll need to replug checkForCommonEnvironmentErrors

dubzzz commented:
By the way, switching to the raw Error object instead of its string versions should drop errors I got reported many times in fast-check that might be related to separateMessageFromStack not able to re-extract the stack and message from some complex and nested ones.

dubzzz commented:
@SimenB Here my drafty work https://github.com/dubzzz/jest/commit/887c2b64dc04e128a4fa750037c79c8532fde14a. I'll rework on it as many things sucks in it. Details being unknown, cause too, I'll need to put something cleaner and safer probably fall-backing onto the original stringified version of the error.

dubzzz commented:
PR opened to add  the support for it and test, see https://github.com/facebook/jest/pull/13935
